src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    public function generateKey(string $secret = null): KeyInterface
+    {
+        $rsa = new RSA();
+        $key = $rsa->createKey();
+
+        if ($key['partialkey']) {
+            throw new EncryptionException('Failed to generate RSA keypair.');
+        }
+
+        if (null === $secret) {
+            $secret = random_bytes(32);
+        }
+
+        return PhpseclibKey::create($secret, $key['privatekey'], $key['publickey']);
+    }
+
+    public function encrypt(string $message, KeyInterface $myKey): string
+    {
+        if (!$myKey instanceof PhpseclibKey) {
+            throw new InvalidKeyException(sprintf('Class "%s" will only accept key objects of class "%s"', self::class, PhpseclibKey::class));
+        }
+
+        set_error_handler(__CLASS__.'::throwError');
+
+        try {
+            $aes = new AES(AES::MODE_CBC);
+            $aes->setPassword($myKey->getSecret());
+            $aes->setIV($nonce = Random::string($aes->getBlockLength() >> 3));
+
+            return Ciphertext::create('AES-CBC', $aes->encrypt($message), $nonce)->getString();
+        } catch (\ErrorException $exception) {
+            throw new EncryptionException(null, $exception);
+        } finally {
+            restore_error_handler();
+        }
+    }
+
+    public function encryptFor(string $message, KeyInterface $recipientKey): string
+    {
+        if (!$recipientKey instanceof PhpseclibKey) {
+            throw new InvalidKeyException(sprintf('Class "%s" will only accept key objects of class "%s"', self::class, PhpseclibKey::class));
+        }
+
+        try {
+            $rsa = new RSA();
+            $rsa->loadKey($recipientKey->getPublicKey());
+            $rsa->setEncryptionMode(RSA::ENCRYPTION_OAEP);
+
+            return Ciphertext::create('RSA-OAEP', $rsa->encrypt($message), random_bytes(8))->getString();
+        } catch (\ErrorException $exception) {
+            throw new EncryptionException(null, $exception);
+        } finally {
+            restore_error_handler();
+        }
+    }
+
+    public function encryptForAndSign(string $message, KeyInterface $recipientKey, KeyInterface $senderKey): string
+    {
+        if (!$recipientKey instanceof PhpseclibKey || !$senderKey instanceof PhpseclibKey) {
+            throw new InvalidKeyException(sprintf('Class "%s" will only accept key objects of class "%s"', self::class, PhpseclibKey::class));
+        }
+
+        try {
+            $rsa = new RSA();
+            $rsa->loadKey($recipientKey->getPublicKey());
+            $rsa->setEncryptionMode(RSA::ENCRYPTION_OAEP);
+            $ciphertext = $rsa->encrypt($message);
+
+            // Load private key after encryption
+            $rsa->loadKey($senderKey->getPrivateKey());
+            $rsa->setSignatureMode(RSA::SIGNATURE_PSS);
+            $headers['signature'] = base64_encode($rsa->sign($ciphertext));
```

Comment:
Why do we `base64_encode`? This should be the job of `Ciphertext`?

Reply:
json_encode cannot encode non ascii strings. 

So I could loop over `$headers` and run `base64_encode` on all values before I json_encode it... but Im not sure that is great. This seams eaasier.